identity-wallet/src/state/credentials/reducers/vp_token_payload_prep.rs

```diff
@@ -0,0 +1,109 @@
+use crate::error::AppError;
+use chrono::{Duration, Utc};
+use identity_core::common::Object as IotaObject;
+
+use identity_credential::{credential::Jwt, presentation::Presentation};
+use identity_iota::core::Url;
+use identity_iota::did::{CoreDID, DIDUrl, DID};
+use jsonwebtoken::Algorithm;
+use jsonwebtoken::Header;
+use oid4vc::oid4vc_core::authorization_request::{AuthorizationRequest, Object};
+use oid4vc::oid4vc_core::{jwt, Subject};
+use oid4vc::oid4vp::dcql::dcql_query::{CredentialQuery, Format};
+use oid4vc::oid4vp::oid4vp::OID4VP;
+use oid4vc::oid4vp::token::{
+    verifiable_presentation_jwt::VerifiablePresentationJwt,
+    vp_token::{PresentationFormat, VpToken},
+    vp_token_builder::VpTokenBuilder,
+};
+use serde_json::Value;
+use std::sync::Arc;
+
+pub async fn prepare_vp_token_object(
+    selected_verifiable_credentials: Vec<(CredentialQuery, Value)>,
+    subject_did: &CoreDID,
+    subject_manager: &Arc<dyn Subject>,
+    oid4vp_authorization_request: &AuthorizationRequest<Object<OID4VP>>,
+    signing_algorithm: Algorithm,
+) -> Result<VpToken, AppError> {
+    let verifier_audience = oid4vp_authorization_request.body.client_id.to_string();
+    let required_nonce = oid4vp_authorization_request.body.extension.nonce.clone();
+
+    let dcql_query = &oid4vp_authorization_request.body.extension.dcql_query;
+    let mut builder = VpTokenBuilder::builder_dcql_query(dcql_query.clone());
+    let signing_method_id =
+        DIDUrl::parse(format!("{}#{}", subject_did, subject_did.method_id())).map_err(|_| AppError::DidParseError)?;
```

Comment:
The value you're trying to generate here is the `kid` or `key_id` which in the context of DIDs is a DID URL which can be resolved to a Public Key (JWK) in the Holder's DID Document. Different DID methods have different methods of 'formatting' the fragment part of the DID URL (fragment part is the part after the `#` character), so the current solution will not work for all DID methods. 

Fortunately we have an abstraction layer for this via the `Sign` trait which is implemented for `Subject`/`subject_manager`:
```rust
    let did_method = format!("did:{}", subject_did.method());

    let key_id = subject_manager
        .key_id(&did_method, signing_algorithm)
        .await
        .ok_or_else(|| AppError::Error(format!("Failed to get signing method ID for DID method {did_method}")))?;
```